### ISSUE TITLE
data syncing to wrong users

### ISSUE THREAD

**Issue by VagueGit** [Mimetis/Dotmim.Sync#1055]
#598 is now occurring daily for the small number of customers we have migrated to DMS 0.9.8. Our customers are companies. Each company has it's own SQL Server database on our server accessed through our web api. Each company has multiple users. Each user syncs to a local SQLite database. Sync scopes and schema are identical across all databases.
**EDIT:** All users in a company sync the same data. We do not use filters.
**EDIT:** We do not use snapshots.

We have over 700 companies using our product with DMS 0.9.1. We occasionally hear that customers see data from other companies. This usually occurred when a new machine was being added to an existing company and the local SQLite database was being populated by sync from the server ... lots of data being synced. Data seems to cross between companies that are syncing at the same time (ie same time-zone/business hours). Presumably these companies are syncing at the same time.

We have migrated a handful of companies to DMS 0.9.8. These companies are in regions where we do not have many customers in timezones outside our main customer concentrations, so the risk of collisions is lower. The three companies we service in one country have all reported instances of seeing another company's data in the first couple of weeks of using our app with DMS 0.9.8.

The server is .NET 6.0. Setup is
```
public static void Main(string[] args)
{
    WebApplicationBuilder builder = WebApplication.CreateBuilder(args);
    IServiceCollection services = builder.Services;
    services.AddControllersWithViews();
    services.AddRazorPages();
    services.AddDistributedMemoryCache();
    services.AddSession(options => options.IdleTimeout = TimeSpan.FromMinutes(30));

    AddSyncServers(services, builder.Configuration);

    builder.Logging.AddElmahIo(options =>
    {
        string elmahApiKey = builder.Configuration.GetValue<string>("ElmahIo:ApiKey");
        string elmahLogId = builder.Configuration.GetValue<string>("ElmahIo:LogId");
        options.ApiKey = elmahApiKey;
        options.LogId = new Guid(elmahLogId);
    });
    WebApplication app = builder.Build();
    app.UseElmahIoExtensionsLogging();
    app.UseRouting();
    app.UseSession();
    app.UseStaticFiles();

    app.UseEndpoints(endpoints =>
    {
        endpoints.MapDefaultControllerRoute();
        endpoints.MapRazorPages();
    });

    app.Run();
}

static void AddSyncServers(IServiceCollection services, ConfigurationManager config)
{
    SyncOptions syncOptions = new SyncOptionsShared().GetSyncOptionsShared();

    string noDatabaseConnectionString = null;
    Helpers.DatabaseHelper databaseHelper = new(config);
    noDatabaseConnectionString = databaseHelper.DatabaseConnection(database: string.Empty);

    SyncScopes syncScopes = new SyncScopes();

    string scopeName = SyncScopes.All;
    SyncSetup syncSetup = syncScopes.GetSyncSetup(scopeName);
    services.AddSyncServer<SqlSyncChangeTrackingProvider>(connectionString: noDatabaseConnectionString, scopeName: scopeName, setup: syncSetup, options: syncOptions);
}
```
Could the static declaration be the issue? Note the connection string is null. This is set during sync.
```
[Route("api/[controller]")]
[Route("api/[controller]/[action]")]
[ApiController]
public class SyncController : ControllerBase
{
    private IEnumerable<WebServerAgent> webServerAgents;
    private readonly IWebHostEnvironment env;
    private readonly ILogger _logger;
    private readonly IConfiguration _config;

    // Injected thanks to Dependency Injection
    public SyncController(IEnumerable<WebServerAgent> webServerAgents,
                            IWebHostEnvironment env, IConfiguration config, ILogger<ElmahIoLogger> logger)
    {
        _config = config;
        _logger = logger;
        this.webServerAgents = webServerAgents;
        this.env = env;
        this._config = config;
    }

    public async Task Post([FromHeader] string databaseName, [FromHeader] string token)
    { 
        try
        {
            UserAccountHelper userAccountHelper = new UserAccountHelper(_config);
            bool isAuthorised = await userAccountHelper.IsAuthorisedAsync(databaseName, token);
            if (!isAuthorised)
                throw new UnauthorizedAccessException();

            string scopeName = HttpContext.GetScopeName();
            WebServerAgent webServerAgent = webServerAgents.FirstOrDefault(
                    c => c.ScopeName == scopeName);

            DatabaseHelper databaseHelper = new(_config);
            webServerAgent.Provider.ConnectionString = databaseHelper.DatabaseConnection(databaseName);
            await webServerAgent.HandleRequestAsync(HttpContext).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            if (HttpContext is not null)
                ex.Ship(HttpContext);
            throw;
        }
    }
}
// the method that returns the connection string for the request database
public string DatabaseConnection(string database)
{
    SqlConnectionStringBuilder sqlStrBldr = new();
    sqlStrBldr.DataSource = _config.GetValue<string>("Database:DataSource");
    sqlStrBldr.UserID = _config.GetValue<string>("Database:UserID");
    sqlStrBldr.Password = _config.GetValue<string>("Database:Password");
    sqlStrBldr.IntegratedSecurity = false;
    sqlStrBldr.PersistSecurityInfo = false;

    sqlStrBldr.ConnectRetryCount = 1;
    sqlStrBldr.ConnectRetryInterval = 10;
    sqlStrBldr.MultipleActiveResultSets = true;
    sqlStrBldr.TrustServerCertificate = true;
    sqlStrBldr.InitialCatalog = database;
    return sqlStrBldr.ConnectionString;
}

```
The sync agent is configured on the client like this
```
static async Task<SyncAgent> ConfigureSyncAgentAsync(SyncType syncType = SyncType.Normal)
{// switch to row at a time processing on foreign key constraint error
    // /https://github.com/Mimetis/Dotmim.Sync/discussions/783#discussioncomment-3379875
    SyncOptions syncOptions = new SyncOptionsShared().GetSyncOptionsShared();
    syncOptions.DisableConstraintsOnApplyChanges = true; // disable on client only during sync
    SqliteSyncProvider clientProvider = GetSqliteSyncProvider(sync: true);
    WebRemoteOrchestrator webRemoteOrchestrator = GetWebRemoteOrchestrator();
    SyncAgent agent = new SyncAgent(clientProvider, webRemoteOrchestrator, syncOptions);
    ConfigureErrorHanding(agent.LocalOrchestrator);
    //        https://github.com/Mimetis/Dotmim.Sync/issues/922#issuecomment-1369540710
    await ConfigureOutdatedAsync(agent.LocalOrchestrator);
    return agent;
}
```

The remote orchestrator for the sync agent is configured to carry credentials in the request header
```
static WebRemoteOrchestrator GetWebRemoteOrchestrator()
{
    string serviceUri = $"{SyncApi}";
    HttpClient httpClient = UserAccountService.HttpClientAuthorize();
    httpClient.Timeout = TimeSpan.FromMinutes(20);
    WebRemoteOrchestrator webRemoteOrchestrator = new WebRemoteOrchestrator(serviceUri, client: httpClient, maxDownladingDegreeOfParallelism: 4);
    webRemoteOrchestrator.HttpClient.Timeout = TimeSpan.FromMinutes(20);
    //ConfigureErrorLogging(webRemoteOrchestrator);
    return webRemoteOrchestrator;
}
// the client authorised for this database
internal static HttpClient HttpClientAuthorize()
{// server to check if token authorized to access db
    HttpClient httpClient = InternetHelper.HttpClientWithCertificateCustomValidationCallback();
    httpClient.DefaultRequestHeaders.Add(nameof(DataService.DatabaseName), $"{DataService.DatabaseName}");
    httpClient.DefaultRequestHeaders.Add(nameof(token), token);
    return httpClient;
}
```
Please tell me I've done something obviously wrong.

**Comment by VagueGit** [comment]
Could this issue arise because all the companies sync using the default batch directory? Would this issue go away if we did this in SyncController?
```
string dmsDirectory = webServerAgent.Options.BatchDirectory;
string userBatchDirectory = Path.Combine(dmsDirectory, databaseName);
webServerAgent.Options.BatchDirectory = userBatchDirectory;
await webServerAgent.HandleRequestAsync(HttpContext).ConfigureAwait(false);
```

**Comment by Mimetis** [comment]
Hello @VagueGit 
Thanks for the detailed explanations.
Needs a little bit of time to review your issue here. 
I will work on it later this month.

I will look into this batch directory thing you mentioned, that could be a good investigation starting point.

I'll keep you updated

**Comment by VagueGit** [comment]
In the meantime, so I can get my customers syncing again, would it be worth my adding this to the SyncController
```
string dmsDirectory = webServerAgent.Options.BatchDirectory;
string userBatchDirectory = Path.Combine(dmsDirectory, databaseName);
webServerAgent.Options.BatchDirectory = userBatchDirectory;
await webServerAgent.HandleRequestAsync(HttpContext).ConfigureAwait(false);
```

**Comment by VagueGit** [comment]
Thinking this through it might be better in the SyncController to
string tempDirectory = Path.GetTempPath();
string userBatchDirectory = Path.Combine(tempDirectory, databaseName);
webServerAgent.Options.BatchDirectory = userBatchDirectory;

**Comment by VagueGit** [comment]
This change in SyncController didn't help
```
string tempDirectory = Path.GetTempPath();
string userBatchDirectory = Path.Combine(tempDirectory, databaseName);
webServerAgent.Options.BatchDirectory = userBatchDirectory;
```
Users are still seeing data from other companies.

Does DMS use Options.BatchDirectory on the server or is this option ignored on the server?

**Comment by VagueGit** [comment]
I implemented a SetupFilter and get HttpSyncWebException
```
[InternalUpgradeAsync]..Invalid object name 'dbo.scope_info_server'.
  -----------------------

  └  ─ Invalid object name 'dbo.scope_info_server'.
```

**Comment by Mimetis** [comment]
> This change in SyncController didn't help
> 
> ```
> string tempDirectory = Path.GetTempPath();
> string userBatchDirectory = Path.Combine(tempDirectory, databaseName);
> webServerAgent.Options.BatchDirectory = userBatchDirectory;
> ```
> 
> Users are still seeing data from other companies.
> 
> Does DMS use Options.BatchDirectory on the server or is this option ignored on the server?
> 
> **EDIT:** As a workaround I am considering adding a DatabaseName column to the top level of the relational hierarchy and to lookup tables. Then apply a DMS [SetupFilter](https://dotmimsync.readthedocs.io/Configuration.html#filtering-rows) to rows of data on the client and the server to filter on DatabaseName.
> 
> I'd appreciate any feedback on this workaround to prevent users from one company seeing data from another company's server database.

Im working on it

**Comment by Mimetis** [comment]
Thinking about your problem / configurations convinced me that using `scopeName` as a filter for multi syncs scenarios (involving potentially multiple `WebServerAgent` instances on the server side Web API) was not a good idea.

You may have (as you) a lot of server databases, with the same schema (and so the same scope name)
You may need to affect the web server agent "_on the fly_" depending on the client request.

I've changed the way you are registering a web server agent, using a new property called `Identifier`.
This new property is only useful if you need multiple `WebServerAgent` instances. 
The web server agent will not have a `ScopeName` property anymore as the scope name is coming from the client side.

It won't change too much things for you.

Registering a multi database sync scenarios (involving multiple web server agents) is still simple. Just use the `Identifier` property to identify all your instances:

``` cs
public void ConfigureServices(IServiceCollection services)
{
    services.AddControllers();

    services.AddDistributedMemoryCache();
    services.AddSession(options => options.IdleTimeout = TimeSpan.FromMinutes(30));

    // [Required]: Get a connection string to your server data source
    var connectionString = Configuration.GetSection("ConnectionStrings")["SqlConnection"];
    var mysqlConnectionString = Configuration.GetSection("ConnectionStrings")["MySqlConnection"];

    var options = new SyncOptions { };

    // [Required] Tables involved in the sync process:
    var tables = new string[] {"ProductCategory", "ProductModel", "Product",
    "Address", "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail" };

    // [Required]: Add a SqlSyncProvider acting as the server hub.
    services.AddSyncServer<SqlSyncChangeTrackingProvider>(connectionString, tables, options, identifier: "sql");
    services.AddSyncServer<MySqlSyncProvider>(mysqlConnectionString, tables, options, identifier : "mysql");
}

```

On the client side, just add the correct identifier, depending on your need, to the `WebRemoteOrchestrator` instance:

``` cs
var serverOrchestrator = new WebRemoteOrchestrator("https://localhost:44342/api/sync", identifier:"sql");
var clientProvider = new SqlSyncProvider(clientConnectionString);

var options = new SyncOptions
{
    BatchSize = 1000
};

// Creating an agent that will handle all the process
var agent = new SyncAgent(clientProvider, serverOrchestrator, options);

```

Eventually, on the web api controller, just make a quick filter based on the header coming from client:

``` cs
private IEnumerable<WebServerAgent> webserverAgents;

// Injected thanks to Dependency Injection
public SyncController(IEnumerable<WebServerAgent> webServerAgents) 
    => this.webserverAgents = webServerAgents;

/// <summary>
/// This POST handler is mandatory to handle all the sync process
/// </summary>
/// <returns></returns>
[HttpPost]
public async Task Post()
{
    var identifier = HttpContext.GetIdentifier();

    var webserverAgent = webserverAgents.FirstOrDefault(c => c.Identifier == identifier);

    await webserverAgent.HandleRequestAsync(HttpContext).ConfigureAwait(false);

}
```

And that's it.

@VagueGit, for your concern about potential conflicts, and the fact you don't know which connection string you need to use before receiving the client request, I can imagine that you can do everything in the controller.

Remove everything you have from the `ConfigureServices` method (in the `startup.cs` file) and create your `WebServerAgent` instance every time with the correct `SyncOptions`:

``` cs
 public SyncController(IConfiguration configuration)
 {
     this.Configuration = configuration;
 }

 /// <summary>
 /// This POST handler is mandatory to handle all the sync process
 /// </summary>
 /// <returns></returns>
 [HttpPost]
 public async Task Post()
 {
     // Let's assume you have the database name in identifier
     var identifier = HttpContext.GetIdentifier();
     
     // Building the correct connection string 
     var connectionString = GetDatabaseConnection(identifier);

     // Create the server provider
     var serverProvider = new SqlSyncProvider(connectionString);

     // Create the sync options related to this identifier:
     var options = new SyncOptions
     {
         // Batch directory will be something like [User Temp Path]/[DotmimSync]/[Db_NAME]
         BatchDirectory = Path.Combine(SyncOptions.GetDefaultUserBatchDirectory(), identifier),
         DisableConstraintsOnApplyChanges = true,
     };

     var tables = new string[] {"ProductCategory", "ProductModel", "Product",
     "Address", "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail" };

     var webserverAgent = new WebServerAgent(serverProvider, tables, options);

     await webserverAgent.HandleRequestAsync(HttpContext).ConfigureAwait(false);
 }


```

That way you are sure a new instance is created every time, and will point to the correct batch directory.
Make sense ?

**Comment by VagueGit** [comment]
Wow ... thank you. When will this be available in the nuget?

**Comment by Mimetis** [comment]
As soon as tests are good and green

**Comment by Mimetis** [comment]
As soon as tests are good and green

**Comment by Mimetis** [comment]
You can try the `v1.0.`0 prerelease here : https://www.nuget.org/packages/Dotmim.Sync.Core/1.0.0-beta-1489

**Comment by VagueGit** [comment]
Got it thanks. Lots of build errors but that was expected as you explained. I'll let you know how we go.

**Comment by VagueGit** [comment]
It seems it is necessary to add a new sync server to services each time a new company is added. If I understand that correctly, this won't work for us.

New companies download and install our software every day. The process of adding a new database is automated from our web app when a new client installs our software.

Or have I misunderstood?

**Comment by VagueGit** [comment]
Would it be possible to [inject the sync service](https://stackoverflow.com/a/56058346/5528313)? This would allow the implementation to be changed at runtime ... maybe?

**Comment by VagueGit** [comment]
On my first attempt running against LocalHost, the request gets to the first curly brace in Post then exits 
```
public async Task Post([FromHeader] string databaseName, [FromHeader] string loginHash, [FromHeader] string scopeName)
{  // request gets to here then exits
    try
    {
        UserAccountHelper userAccountHelper = new UserAccountHelper(_config);
        bool isAuthorised = await userAccountHelper.IsAuthorisedAsync(databaseName, loginHash);
        if (!isAuthorised)
            throw new UnauthorizedAccessException();

        DatabaseHelper databaseHelper = new(_config);
        databaseName = HttpContext.GetIdentifier();
        string connectionString = databaseHelper.DatabaseConnection(databaseName);
        SqlSyncChangeTrackingProvider serverProvider = new SqlSyncChangeTrackingProvider(connectionString);

        SyncSetup syncSetup = SyncHelper.GetSyncSetup(scopeName);
        SyncOptions syncOptions = SyncHelper.GetSyncOptionsShared(databaseName);
        WebServerAgent webServerAgent = new(serverProvider, syncSetup, syncOptions);

        await webServerAgent.HandleRequestAsync(HttpContext).ConfigureAwait(false);
    }
    catch (Exception ex)
    {
        if (HttpContext is not null)
            ex.Ship(HttpContext);
        throw;
    }
}

```
A SyncException is thrown on the client 
```
try
{
    syncResult = await syncAgent.SynchronizeAsync(scopeName: scopeName, setup: syncSetup, syncType: syncType, progress: progress);
}
catch (HttpSyncWebException ex) 
{
}
catch (SyncException ex) // I get 'column is null' here
```
Note it is SyncException caught, not HttpSyncWebException

On the client we add the identifier
```
WebRemoteOrchestrator webRemoteOrchestrator = new WebRemoteOrchestrator(
    serviceUri: serviceUri,
    client: httpClient,
    maxDownladingDegreeOfParallelism: 4,
    identifier: DataService.DatabaseName);
```
Not much in the stack trace
```
   at Dotmim.Sync.SyncAgent.<SynchronizeAsync>d__28.MoveNext()
   at Wbl.Services.DataSyncService.<SyncAsync>d__20.MoveNext() in C:\Files ...
```

**Comment by VagueGit** [comment]
The exception might be thrown by the SyncController? So I've trimmed down the ctor for the SyncController like this
```
public class SyncController : ControllerBase
{
    //private IEnumerable<WebServerAgent> webServerAgents;
    //private readonly IWebHostEnvironment env;
    //private readonly ILogger _logger;
    private readonly IConfiguration _config;

    // Injected thanks to Dependency Injection
    public SyncController(IConfiguration config)
    {
        _config = config;
        //_logger = logger;
        //this.webServerAgents = webServerAgents;
        //this.env = env;
    }
```
The stack trace from the inner exception shows
```
   at Dotmim.Sync.SyncRow.get_Item(String columnName)
   at Dotmim.Sync.LocalOrchestrator.<UpgdrateFromNoWhereTo098Async>d__32.MoveNext()
   at Dotmim.Sync.LocalOrchestrator.<UpgdrateFromNoWhereTo098Async>d__32.MoveNext()
   at Dotmim.Sync.LocalOrchestrator.<InternalUpgradeAsync>d__31.MoveNext()
   at Dotmim.Sync.LocalOrchestrator.<InternalUpgradeAsync>d__31.MoveNext()
   at Dotmim.Sync.SyncAgent.<SynchronizeAsync>d__28.MoveNext()
```

**Comment by Mimetis** [comment]
Oh, I think the auto upgrade routine is not working correctly......

I'm publishing a new version that will fix this bug.

Otherwise you can update the version on each side (I think only client side is impacted) manually by running `UPDATE scope_info SET sync_scope_version = '1.0.0'` on both side.

**Comment by VagueGit** [comment]
Thank you. The workaround now has sync functioning but not updating.

Testing sync a single table to keep the last job number which only allows updates, so it only ever has a single row.

The problem seems to be that DMS tries to set DbName to null. The DbName column has been added to prevent data from wrong company.

I am reluctant to remove the DbName column until we prove this fix in the wild. Also DbName cannot be removed from the local db as SQLite does not allow ALTER TABLE.

There are no setup filters. All columns should be synced.

On the client SQLite db 
```
CREATE TABLE JobNrs (
Id INTEGER NOT NULL DEFAULT 0, 
Last INTEGER NOT NULL, 
DbName TEXT NOT NULL DEFAULT NZ16446 CHECK (DbName='MyTestDb'), 
PRIMARY KEY(Id));
```
On the sql server
```
CREATE TABLE [dbo].[JobNrs](
	[Id] [int] NOT NULL,
	[Last] [int] NOT NULL,
	[DbName] [nvarchar](14) NOT NULL,
 CONSTRAINT [PK_JobNrs] PRIMARY KEY CLUSTERED ([Id] ASC)
)
GO
ALTER TABLE [dbo].[JobNrs] ADD  CONSTRAINT [DF_JobNrs_Id]  DEFAULT ((0)) FOR [Id]
GO
ALTER TABLE [dbo].[JobNrs] ADD  DEFAULT (db_name()) FOR [DbName]
GO
ALTER TABLE [dbo].[JobNrs]  WITH CHECK ADD  CONSTRAINT [CHK_JobNrs_DbName] CHECK  (([DbName]=db_name()))
GO
ALTER TABLE [dbo].[JobNrs] CHECK CONSTRAINT [CHK_JobNrs_DbName]
GO

```
On either the client or the server I can update Last with
`UPDATE JobNrs SET Last=3`
However I get this from orchestrator.OnApplyChangesErrorOccured
```
SchemaTable: JobNrs
ApplyType: Modified
ErrorRow: [Sync state]:Modified, [Id]:0, [Last]:3, [DbName]:<NULL />
Message: SQLite Error 19: 'NOT NULL constraint failed: JobNrs.DbName'.
```
Why would DMS try to null DBName?

**Comment by Mimetis** [comment]
Did you create a new SyncSetup for this table ? 
Otherwise, need a sample to reproduce

**Comment by VagueGit** [comment]
The client is a new install but connects to an existing server database. The server was deprovisioned like this
`DROP TABLE IF EXISTS scope_info, scope_info_history, scope_info_server`
Then reprovisioned by DMS on first sync

This is the sync_scope_schema on the client
`{"t":[{"n":"JobNrs","s":"","op":"SqlSyncProvider, Dotmim.Sync.SqlServer.SqlSyncProvider","c":[{"n":"Id","dt":"6","iu":true,"ps":true,"p1":10,"odb":"Int","oty":"int","db":11,"dv":"((0))"},{"n":"Last","dt":"6","o":1,"ps":true,"p1":10,"odb":"Int","oty":"int","db":11},{"n":"DbName","dt":"17","iuc":true,"ml":14,"o":2,"odb":"NVarChar","oty":"nvarchar","db":16,"dv":"(db_name())"}],"pk":["Id"]}],"r":[],"f":[]}
`

This is scope_info on the server
`{"t":[{"n":"JobNrs","s":"","op":"SqlSyncProvider, Dotmim.Sync.SqlServer.SqlSyncProvider","c":[{"n":"Id","dt":"6","iu":true,"ps":true,"p1":10,"odb":"Int","oty":"int","db":11,"dv":"((0))"},{"n":"Last","dt":"6","o":1,"ps":true,"p1":10,"odb":"Int","oty":"int","db":11},{"n":"DbName","dt":"17","iuc":true,"ml":14,"o":2,"odb":"NVarChar","oty":"nvarchar","db":16,"dv":"(db_name())"}],"pk":["Id"]}],"r":[],"f":[]}`

Do they look ok?

**Comment by Mimetis** [comment]
Dont know, seems good to me
Need a sample to reproduce

**Comment by VagueGit** [comment]
hmmm ... I appreciate your help @Mimetis but I'm not sure how to unpick this issue from a quite large app.

The DbName table was only added to deal with data being synced to the wrong companies so is redundant for business purposes.

I've looked at reprovisioning but I'm struggling to get my head around [ScopeInfoClients](https://dotmimsync.readthedocs.io/ScopeClients.html?highlight=scope_info_client#scopeinfoclients)

> A scope client is a the association of one scope with a filter, and is stored in the scope_info_client table.

We don't use filters, yet for the JobNrs scope there are 44 rows in scope_info_client on the server.

If we assume the problem of data going to the wrong company is now fixed, would an interceptor allow us to catch any dbName=null being synced and assign the correct value? If so, which interceptor would you recommend?

Once clients have uploaded unsynced data they can uninstall the app and reinstall a version that doesn't use DbName.

**Comment by Mimetis** [comment]
Honestly I don't know, not sure an interceptor will help here.

ScopeInfoClient represents a client database. If you don't have filters, each line is representing one client

**Comment by VagueGit** [comment]
The docs on [Intercepting rows](https://dotmimsync.readthedocs.io/Interceptors.html#intercepting-rows)

> Or even intercept all the rows that are going to be applied on a destination database.
> That way, you may be able to modify these rows, to meet your business / requirements rules.

I thought this might allow us to intercept DbName=null and fix it. The interceptor examples in the docs all write to the console but  I can't find an example that shows how to edit data. That might get us out of a jam. Are you able to give an example of how to intercept and change all rows where DMS tries to set DbName=null?

**Comment by Mimetis** [comment]
You are trying to fix something that should not happens.
I think it's not a good idea :)

**Comment by VagueGit** [comment]
I agree absolutely but it is happening and companies are losing money and laying off staff because they are unable to process orders.

We need to patch this temporarily to get them back in business.

**Comment by Mimetis** [comment]
good luck !

**Comment by VagueGit** [comment]
Tell that to people with kids and a mortgage

**Comment by Mimetis** [comment]
Is it some kind of pressure / harassment you try to put on me ?

**Comment by Mimetis** [comment]
As you say, we are not leaving in a perfect world.

I work on this project **FOR FREE** on my **SPARE TIME**, and I don't have currently enough time to resolve **YOUR** issue.
If the project does not fit your needs and is, in your opinion, too buggy, feel free to try another solution.

For your problem, and once again, it's not a solution in my opinion, but you can play with https://dotmimsync.readthedocs.io/Interceptors.html#onrowschangesapplying

``` cs
agent.LocalOrchestrator.OnRowsChangesApplying(rca =>
{
    foreach(var row in rca.SyncRows)
    {
        // change row here, before updating / inserting in local store
        row["DbName"] = "dddd";
    }
});
```

And yes, it's part of the documentation

**Comment by VagueGit** [comment]
I've got DMS 1.0.0-beta-1489 working with a new database (client and server) but fails on an existing database with `HttpSyncWebException`
```
InternalUpgradeAsync]..Invalid object name 'dbo.scope_info_server'.
  -----------------------
  └  ─ Invalid object name 'dbo.scope_info_server'.
```
Those dbs were using DMS 0.9.4 before upgrading to DMS 1.0.0-beta-1489

**Comment by Mimetis** [comment]
Once again, you are not helping here.
No information on your setup and the action you are doing to end up with this issue.

- I don't know on which side the bug is occurring
- I don't know if you have upgraded only one side or the two sides to the last version
- I don't know why you are trying to remove the tables `scope_info`, `scope_info_history`, `scope_info_server `(by the way, the tables have been replaced by `scope_info`, `scope_info_clients `only)
- I don't know on which side you are removing them
- I don't know why one of your field is not sync correctly (`DBName`) 
- I don't know if your databases are provisioned correctly
- I don't know how your are provisioning the new databases
- I don't know everything else that can help to resolve your issue

Regarding the discussion, I've tried several scenarios, involving dropping the sync_* tables, and all my tests are good.
Sorry, can't help you here until you are providing a good example and a way to reproduce it.
